File: src/renderer/lib/type/block.ts
type Position = { x: number, y: number };

type ValueContent = {
    title: string;
    value: string;
    placeholder: string;
}

type SeparatorContent = {
   type: 'line' | 'space';
}

type SelectContent = {
    title: string;
    options: string[];
    value: string;
}

interface BaseBlockContent<T,C> {
    id: string;
    type: T;
    content: C;
}

interface ValueBlockContent extends BaseBlockContent<'value', ValueContent> {
    id: string;
    type: 'value';
    content: ValueContent;
}

interface SeparatorBlockContent extends BaseBlockContent<'separator', SeparatorContent> {
    id: string;
    type: 'separator';
    content: SeparatorContent;
}

interface SelectBlockContent extends BaseBlockContent<'select', SelectContent> {
    id: string;
    type: 'select';
    content: SelectContent;
}

type BlockContent = ValueBlockContent | SeparatorBlockContent | SelectBlockContent;

interface BlockType {
    id: string;
    type: 'flag' | ('move' | 'composition' | 'works') | 'loop' | 'value';
    title: string;
    output: string;
    position: Position;
    connections: {
        input: Position | null;
        output: Position | null;
    };
    contents: BlockContent[];
    childId: string;
    parentId: string;
    depth: number;
    zIndex: number;
}

export type {
    Position,
    BlockContent,
    BlockType
}

File: src/models/block.rs
#[derive(Clone, Debug, PartialEq)]
pub struct Position {
    pub x: f64,
    pub y: f64,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Size {
    pub width: f64,
    pub height: f64,
}

#[derive(Clone, Debug, PartialEq)]
pub enum Connection {
    Input,
    Output,
    Both,
    None,
}

#[derive(Clone, Debug, PartialEq)]
pub enum BlockType {
    Flag,
    Works,
    Move,
    Composition,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ContentValue {
    pub title: String,
    pub value: String,
    pub placeholder: Option<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum Separator {
    None,
    Space,
    Newline,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Options {
    pub id: String,
    pub title: String,
    pub value: String,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ContentSelector {
    pub title: String,
    pub value: String,
    pub options: Vec<Options>,
    pub placeholder: Option<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum EnumBlockContent {
    ContentValue(ContentValue),
    ContentSelector(ContentSelector),
    Separator(Separator),
}

#[derive(Clone, Debug, PartialEq)]
pub struct BlockContent {
    pub id: String,
    pub content: EnumBlockContent,
}

#[derive(Clone, PartialEq)]
pub struct Block {
    pub id: String,
    pub block_type: BlockType,
    pub title: String,
    pub output: String,
    pub content: Vec<BlockContent>,
    pub connection: Connection,
    pub child_id: Option<String>,
    pub parent_id: Option<String>,
    pub position: Position,
    pub size: Size,
    pub z_index: i32,
}

impl Block {
    pub fn new(
        id: String,
        block_type: BlockType,
        title: String,
        position: Position,
    ) -> Self {
        Self {
            id,
            block_type,
            title,
            output: String::new(),
            content: Vec::new(),
            connection: Connection::None,
            child_id: None,
            parent_id: None,
            position,
            size: Size { width: 150.0, height: 60.0 },
            z_index: 0,
        }
    }
}
